pkg/store/bucket.go

```diff
@@ -661,10 +700,17 @@ func debugFoundBlockSetOverview(logger log.Logger, mint, maxt int64, lset labels
 }
 
 // Series implements the storepb.StoreServer interface.
-// TODO(bwplotka): It buffers all chunks in memory and only then streams to client.
-// 1. Either count chunk sizes and error out too big query.
-// 2. Stream posting -> series -> chunk all together.
 func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_SeriesServer) error {
+	{
+		span, _ := tracing.StartSpan(srv.Context(), "bucket_store_gate")
+		err := s.queryGate.IsMyTurn(srv.Context())
+		span.Finish()
+		if err != nil {
+			return errors.Wrapf(err, "failed to wait for turn")
+		}
+		defer s.queryGate.Done()
```

Comment:
I think you want to actually do this. As defer will end until the whole function is done?

Reply:
But then the counter would go down at exactly this point and then another query could come through even though the old one is still being parsed. Or am I wrong?

Reply:
ignore me, I confused this with `span.finish`. Can you at least put this defer outside of code block? It is quite confusing as at first glance it shows the intent of this defer to be end at `}`. (:

Reply:
Will do, missed that one